sonarts-core/src/se/programStates.ts

```diff
@@ -0,0 +1,147 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as ts from "typescript";
+import { SymbolicValue, isEqualSymbolicValues } from "./symbolicValues";
+import { inspect } from "util";
+import {
+  Constraint,
+  getTruthyConstraint,
+  getFalsyConstraint,
+  isEqualConstraints,
+  addConstraintToList,
+} from "./constraints";
+
+type SymbolicValues = Map<ts.Symbol, SymbolicValue>;
+type ExpressionStack = SymbolicValue[];
+type Constraints = Map<SymbolicValue, Constraint[]>;
+
+export class ProgramState {
+  private readonly symbolicValues: SymbolicValues;
+  private readonly expressionStack: ExpressionStack;
+  private readonly constraints: Constraints;
+
+  public static empty() {
+    return new ProgramState(new Map(), [], new Map());
+  }
+
+  private constructor(symbolicValues: SymbolicValues, expressionStack: ExpressionStack, constraints: Constraints) {
+    this.symbolicValues = symbolicValues;
+    this.expressionStack = expressionStack;
+    this.constraints = constraints;
+  }
+
+  sv(symbol: ts.Symbol): SymbolicValue | undefined {
+    return this.symbolicValues.get(symbol);
+  }
+
+  setSV(symbol: ts.Symbol, sv: SymbolicValue) {
+    const newSymbolicValues = new Map(this.symbolicValues);
+    newSymbolicValues.set(symbol, sv);
+    return new ProgramState(newSymbolicValues, this.expressionStack, this.constraints);
+  }
+
+  pushSV(sv: SymbolicValue): ProgramState {
+    const newExpressionStack = [...this.expressionStack, sv];
+    return new ProgramState(this.symbolicValues, newExpressionStack, this.constraints);
+  }
+
+  popSV(): [SymbolicValue | undefined, ProgramState] {
+    const newExpressionStack = [...this.expressionStack];
+    return [newExpressionStack.pop(), new ProgramState(this.symbolicValues, newExpressionStack, this.constraints)];
+  }
+
+  getStackSize() {
+    return this.expressionStack.length;
+  }
+
+  addConstraint(constraint: Constraint) {
+    if (this.expressionStack.length > 0) {
+      const sv = this.expressionStack[this.expressionStack.length - 1];
+      const newConstraints = new Map(this.constraints);
+      const svConstraints = newConstraints.get(sv) || [];
+      newConstraints.set(sv, addConstraintToList(constraint, svConstraints));
+      return new ProgramState(this.symbolicValues, this.expressionStack, newConstraints);
+    } else {
+      throw new Error("Cannot apply a constraint, because the expression stack is empty");
+    }
+  }
+
+  addTruthyConstraint() {
+    return this.addConstraint(getTruthyConstraint());
+  }
+
+  addFalsyConstraint() {
+    return this.addConstraint(getFalsyConstraint());
+  }
+
+  getConstraints(sv: SymbolicValue) {
+    return this.constraints.get(sv) || [];
+  }
+
+  hasEmptyStack(): boolean {
+    return this.expressionStack.length == 0;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') == should be === [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1440)